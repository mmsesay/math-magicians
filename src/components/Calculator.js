import React, { useState, useEffect } from 'react';
import calculate from '../logic/calculate';

export const Calculator = () => {
  const [firstValue, setFirstValue] = useState('');
  const [secondValue, setSecondValue] = useState('');
  const [operator, setOperator] = useState('');
  const [dataObject, setDataObject] = useState();
  const [result, setResult] = useState('');
  const [isCalculate, setIsCalculate] = useState(false);

  const operators = '+-xรท%';
  const numbers = '0123456789';

  const handleValues = (event) => {
    if (operators.includes(event.target.name)) {
      if (firstValue === '') {
        setFirstValue(event.target.name);
      } else {
        setOperator(event.target.name);
      }
      setResult(`${result}${event.target.name}`);
    } 
    
    if (numbers.includes(event.target.name)) {
      // validate for an operation
      if (dataObject.operation === '') {
        // set the firstValue if we have non else concatenate the existing with new value
        if (firstValue === '') {
          setFirstValue(event.target.name);
        } else if (firstValue) {
          setFirstValue(`${firstValue}${event.target.name}`);
        }
      } else {
        setSecondValue(event.target.name);
      }
      setResult(`${result}${event.target.name}`);
    }

    // reset the result and force reload the browser
    if (event.target.name === 'AC') {
      setResult('');
      window.location.reload(true);
    }

    if (isCalculate) {
      const response = calculate(dataObject, event.target.name);
      if (response) {
        // set the total result if both next and operation are null
        if (response.next === null && response.operation === null) {
          setResult(`${response.total || ''}`);
        }

        // validate we have all null response, reset the result and isCalculate  
        if (!response.total && !response.next && !response.operation) {
          setResult('');
          setIsCalculate(false);
        }
        
        // check for dot in the next value and display the exact value
        if (response.next) {
          setResult(`${response.next || ''}`);
        }

        // update the state of the values
        setFirstValue(response.total);
        setSecondValue(response.next);
        setOperator(response.operation);
      }
    }
  };

  useEffect(() => {
    // create a new data object
    const newDataObject = {
      total: firstValue,
      next: secondValue,
      operation: operator,
    };

    // update the data object state
    setDataObject(newDataObject);

    // validate the object have valid data and update the isCalculate state to perfom it calculation
    if (newDataObject.total && newDataObject.next && newDataObject.operation) {
      setIsCalculate(true);
    }
  }, [firstValue, secondValue, operator, isCalculate]);

  return (
    <div className="bg-white w-2/6 h-2/4 rounded oveflow-hidden">
      <div className="bg-gray-500 h-1/6 flex items-center justify-end p-5">
        <p className="text-2xl text-white">{ result }</p>
      </div>
      <div className="bg-gray-400 h-5/6 grid grid-cols-4 gap-1">
        <button name="AC" onClick={handleValues} className="bg-white w-22">AC</button>
        <button name="+/-" onClick={handleValues} className="bg-white w-22">+/-</button>
        <button name="%" onClick={handleValues} className="bg-white w-22">%</button>
        <button name="รท" onClick={handleValues} className="bg-yellow-500 w-22">รท</button>
        <button name="7" onClick={handleValues} className="bg-white w-22">7</button>
        <button name="8" onClick={handleValues} className="bg-white w-22">8</button>
        <button name="9" onClick={handleValues} className="bg-white w-22">9</button>
        <button name="x" onClick={handleValues} className="bg-yellow-500 w-22">x</button>
        <button name="4" onClick={handleValues} className="bg-white w-22">4</button>
        <button name="5" onClick={handleValues} className="bg-white w-22">5</button>
        <button name="6" onClick={handleValues} className="bg-white w-22">6</button>
        <button name="-" onClick={handleValues} className="bg-yellow-500 w-22">-</button>
        <button name="1" onClick={handleValues} className="bg-white w-22">1</button>
        <button name="2" onClick={handleValues} className="bg-white w-22">2</button>
        <button name="3" onClick={handleValues} className="bg-white w-22">3</button>
        <button name="+" onClick={handleValues} className="bg-yellow-500 w-22">+</button>
        <button name="0" onClick={handleValues} className="bg-white w-22 col-span-2">0</button>
        <button name="." onClick={handleValues} className="bg-white w-22">.</button>
        <button name="=" onClick={handleValues} className="bg-yellow-500 w-22">=</button>
      </div>
    </div>
  );
};