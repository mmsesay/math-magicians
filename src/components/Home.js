export const Home = () => (
  <div className="w-full h-auto my-24 py-10 px-20 space-y-10 text-white text-justify">
    <h1 className="text-4xl font-bold mb-20">Welcome to our page!</h1>
    <p className="text-2xl">
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi ultrices, 
      purus sit amet lobortis egestas, libero enim rhoncus libero, in posuere dui velit at libero. 
      Sed non tellus tortor. Maecenas euismod tortor ex, eu placerat odio egestas quis. 
      Proin dignissim elit quam, nec facilisis quam laoreet a. Donec in dignissim erat, 
      sit amet elementum lectus. Donec maximus est vulputate, aliquet eros nec, finibus velit. 
      Nulla at lacus accumsan, euismod diam ac, pretium libero.
    </p>
    <p className="text-2xl">
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi ultrices, 
      purus sit amet lobortis egestas, libero enim rhoncus libero, in posuere dui velit at libero. 
      Sed non tellus tortor. Maecenas euismod tortor ex, eu placerat odio egestas quis. 
      Proin dignissim elit quam, nec facilisis quam laoreet a. Donec in dignissim erat, 
      sit amet elementum lectus. Donec maximus est vulputate, aliquet eros nec, finibus velit. 
      Nulla at lacus accumsan, euismod diam ac, pretium libero.
    </p>
  </div>
); 
