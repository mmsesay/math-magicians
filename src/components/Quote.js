export const Quote = () => (
  <div className="w-full h-auto my-24 py-10 px-20 text-justify">
    <p className="text-2xl text-gray-200 mx-32">
      Mathematics is not about numbers, equations, computations, or algorithms: 
      it is about understanding. - William Paul Thurston
    </p>
  </div>
); 
