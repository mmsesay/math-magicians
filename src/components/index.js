import { Home } from './Home';
import { Calculator } from './Calculator';
import { Quote } from './Quote';

export {
  Home,
  Calculator,
  Quote,
};