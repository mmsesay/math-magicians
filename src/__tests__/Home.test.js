import { render, screen } from '@testing-library/react';
import { Home } from '../components';

describe('Testing homepage', () => {
  test('Tests the existence of the the title: Welcome to our page!', () => {
    render(<Home />);
    const h1 = screen.getByText(/Welcome to our page!/);
    expect(h1).toBeInTheDocument();
  });

  test('Tests the existence of the first paragraph', () => {
    render(<Home />);
    const paragraphs = screen.queryAllByTestId('paragraph');
    expect(paragraphs).toHaveLength(2);
  });
});