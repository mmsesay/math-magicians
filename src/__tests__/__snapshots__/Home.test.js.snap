// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing homepage renders as expected 1`] = `
<div
  className="w-full h-auto my-24 py-10 px-20 space-y-10 text-justify"
>
  <h1
    className="text-4xl text-gray-300 font-bold mb-20"
  >
    Welcome to our page!
  </h1>
  <p
    className="text-2xl text-gray-200"
    data-testid="paragraph"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi ultrices, purus sit amet lobortis egestas, libero enim rhoncus libero, in posuere dui velit at libero. Sed non tellus tortor. Maecenas euismod tortor ex, eu placerat odio egestas quis. Proin dignissim elit quam, nec facilisis quam laoreet a. Donec in dignissim erat, sit amet elementum lectus. Donec maximus est vulputate, aliquet eros nec, finibus velit. Nulla at lacus accumsan, euismod diam ac, pretium libero.
  </p>
  <p
    className="text-2xl text-gray-200"
    data-testid="paragraph"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi ultrices, purus sit amet lobortis egestas, libero enim rhoncus libero, in posuere dui velit at libero. Sed non tellus tortor. Maecenas euismod tortor ex, eu placerat odio egestas quis. Proin dignissim elit quam, nec facilisis quam laoreet a. Donec in dignissim erat, sit amet elementum lectus. Donec maximus est vulputate, aliquet eros nec, finibus velit. Nulla at lacus accumsan, euismod diam ac, pretium libero.
  </p>
</div>
`;
